### PR TITLE
Add mana structure and fix max mana checks

### LINKED ISSUE #1133: Introduce ManaStructure
https://github.com/iotaledger/tips-draft/pull/25/commits/8bf8579b5f8c6a8aa4cfc18279116311b49b578b

### LINKED ISSUE #1148: Rename manaBitsExponent to manaBitsCount
https://github.com/iotaledger/tips-draft/commit/ff9e4e1e8c0128644906a213b31067516e3bdcb7
https://github.com/iotaledger/tips-draft/pull/25/files/8bf8579b5f8c6a8aa4cfc18279116311b49b578b..21d36e35a610b765b3e6e3b510c68bf08e808a92

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

This PR adds the `ManaStructure` to the `ProtocolParameters` and fixes max mana checks.

## Links to any relevant issues

Closes #1133
Closes #1148

### CODE REVIEW

**Review by Thoralf-M — DISMISSED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Need to change the test block ID

File: sdk/src/types/block/protocol.rs [hunk 1]
```diff
@@ -320,6 +295,65 @@ impl Default for VersionSignalingParameters {
...
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(rename_all = "camelCase")
+)]
+#[packable(unpack_error = Error)]
+#[getset(get_copy = "pub")]
+pub struct ManaStructure {
+    /// The number of bits used to represent Mana expressed as an exponent of 2.
+    pub(crate) mana_bits_exponent: u8,
```
Comment (Thoralf-M) [inline sdk/src/types/block/protocol.rs]: Now `mana_bits_count` https://github.com/iotaledger/tips-draft/pull/25/files/8bf8579b5f8c6a8aa4cfc18279116311b49b578b..21d36e35a610b765b3e6e3b510c68bf08e808a92

File: sdk/src/types/block/protocol.rs [hunk 2]
```diff
@@ -320,6 +295,65 @@ impl Default for VersionSignalingParameters {
...
 
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Packable, CopyGetters)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(rename_all = "camelCase")
+)]
+#[packable(unpack_error = Error)]
+#[getset(get_copy = "pub")]
+pub struct ManaStructure {
```
Comment (thibault-martinez) [inline sdk/src/types/block/protocol.rs]: Can you move this to the mana module please?

File: sdk/src/types/block/mana/mod.rs [hunk 3]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-mod allotment;
+pub mod allotment;
```
Comment (thibault-martinez) [inline sdk/src/types/block/mana/mod.rs]: If ever you want to keep it consistent, we usually keep it private and re-export what's important from it

File: sdk/src/types/block/protocol.rs (hunk 1 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/protocol.rs]: ffs

File: sdk/src/types/block/protocol.rs (hunk 2 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/protocol.rs]: sure

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/types/block/mana/protocol.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        self.mana_decay_factors.get(*epoch_index as usize).copied()
+    }
+
+    /// Returns the max mana that can exist with the mana bits defined.
+    pub fn max_mana(&self) -> u64 {
+        (1 << self.mana_bits_count) - 1
+    }
+}
+
+impl Default for ManaStructure {
```
Comment (kwek20) [inline sdk/src/types/block/mana/protocol.rs:54]: Do we need a default?

File: sdk/src/types/block/protocol.rs [hunk 5]
```diff
@@ -160,14 +138,9 @@ impl ProtocolParameters {
...
-    /// Returns the mana decay factors slice of the [`ProtocolParameters`].
-    pub fn mana_decay_factors(&self) -> &[u32] {
-        &self.mana_decay_factors
-    }
-
-    /// Returns the mana decay factor for the given epoch index.
-    pub fn mana_decay_factor_at(&self, epoch_index: EpochIndex) -> Option<u32> {
-        self.mana_decay_factors.get(*epoch_index as usize).copied()
+    /// Returns the parameters used for mana calculations.
+    pub fn mana_structure(&self) -> &ManaStructure {
```
Comment (kwek20) [inline sdk/src/types/block/protocol.rs:142]: dont we get this from the #[getset(get_copy = "pub")] ?

Comment (kwek20) [reply sdk/src/types/block/protocol.rs:142]: otherwise, i htink were missing a few getters in ManaStructure

Comment (DaughterOfMars) [reply sdk/src/types/block/protocol.rs:142]: No, the type is not `Copy` so it is skipped.

**Review by kwek20 — DISMISSED** [review]

File: sdk/src/types/block/mana/allotment.rs [hunk 6]
```diff
@@ -34,8 +31,8 @@ impl Ord for ManaAllotment {
 }
 
 impl ManaAllotment {
-    pub fn new(account_id: AccountId, mana: u64) -> Result<Self, Error> {
-        if mana > THEORETICAL_MANA_MAX {
+    pub fn new(account_id: AccountId, mana: u64, protocol_params: &ProtocolParameters) -> Result<Self, Error> {
+        if mana > protocol_params.mana_structure().max_mana() {
```
Comment (thibault-martinez) [inline sdk/src/types/block/mana/allotment.rs:35]: Discussion for later as we have an issue for that but I'd call it mana_max for consistency

File: sdk/src/types/block/mana/mod.rs [hunk 7]
```diff
@@ -65,27 +64,43 @@ impl ManaAllotments {
...
 
-fn verify_allotments_sum<'a>(allotments: impl IntoIterator<Item = &'a ManaAllotment>) -> Result<(), Error> {
+fn verify_allotments_unique_sorted<'a>(allotments: impl IntoIterator<Item = &'a ManaAllotment>) -> Result<(), Error> {
+    if !is_unique_sorted(allotments.into_iter()) {
+        return Err(Error::ManaAllotmentsNotUniqueSorted);
+    }
+    Ok(())
+}
+
+pub(crate) fn verify_allotments_sum<'a>(
```
Comment (thibault-martinez) [inline sdk/src/types/block/mana/mod.rs]: I forgot to rename these, do you ming adding mana please. Even more important if they are not private anymore

Comment (DaughterOfMars) [reply sdk/src/types/block/mana/mod.rs]: sure

**Review by thibault-martinez — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Review by abdulmth — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Pull request merged**